interface/ceed-qfunctioncontext.c

```diff
@@ -19,10 +19,106 @@
 #include <ceed-impl.h>
 #include <stdint.h>
 #include <stdio.h>
+#include <string.h>
 
 /// @file
 /// Implementation of public CeedQFunctionContext interfaces
 
+/// ----------------------------------------------------------------------------
+/// CeedQFunctionContext Library Internal Functions
+/// ----------------------------------------------------------------------------
+/// @addtogroup CeedQFunctionDeveloper
+/// @{
+
+/**
+  @brief Get index for QFunctionContext field
+
+  @param ctx         CeedQFunctionContext
+  @param field_name  Name of field
+  @param field_index Index of field, or -1 if field is not registered
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref Developer
+**/
+int CeedQFunctionContextGetFieldIndex(CeedQFunctionContext ctx,
+                                      const char *field_name, CeedInt *field_index) {
+  *field_index = -1;
+  for (CeedInt i=0; i<ctx->num_fields; i++)
+    if (!strcmp(ctx->field_descriptions[i].name, field_name))
+      *field_index = i;
+  return CEED_ERROR_SUCCESS;
+}
+
+/**
+  @brief Common function for registering QFunctionContext fields
+
+  @param ctx               CeedQFunctionContext
+  @param field_name        Name of field to register
+  @param field_offset      Offset of field to register
+  @param field_description Description of field, or NULL for none
+  @param field_type        Field data type, such as double or int32
+  @param field_size        Size of field, in bytes
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref Developer
+**/
+int CeedQFunctionContextRegisterGeneric(CeedQFunctionContext ctx,
+                                        const char *field_name, size_t field_offset,
+                                        const char *field_description,
+                                        const char *field_type, size_t field_size) {
+  int ierr;
+
+  // Check for duplicate
+  CeedInt field_index = -1;
+  ierr = CeedQFunctionContextGetFieldIndex(ctx, field_name, &field_index);
+  CeedChk(ierr);
+  if (field_index != -1)
+    // LCOV_EXCL_START
+    return CeedError(ctx->ceed, CEED_ERROR_UNSUPPORTED,
+                     "QFunctionContext field with name \"%s\" already registered",
+                     field_name);
+  // LCOV_EXCL_STOP
+
+  // Allocate space for field data
+  if (ctx->num_fields == 0) {
+    ierr = CeedCalloc(1, &ctx->field_descriptions); CeedChk(ierr);
+  } else {
+    ierr = CeedRealloc(ctx->num_fields + 1, &ctx->field_descriptions);
```

Comment:
Likely no big deal here, but this becomes quadratic complexity if there are many fields so usually one wants to separate the amount of space allocated (which might double each time it's exceeded) from the amount used.